README
 	App for CloudStack API

 	Current version supports followings,
 	*multi cloud services
	 Tata,Ninefold,Datapipe,NTTCom,IDCF,Contegix
	*You can add your CloudStack.
 	*support commands
 	 listVirtulMachines,deployVirtualMachines,destroyVirtualmachines
 	 listVolumes 
 
 	If you use previous version(older than 1.3.0),please delete it and re-install latest version.
 
 
    Copyright (c) 2012, Shugo Numano(snumano@cpan.org)
    All rights reserved.

    Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

        Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
        Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
        Neither the name of the <ORGANIZATION> nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



CHANGELOG.txt
1.3.2   2012/06/18
		Add map.
		Add App version in Info window.

1.3.1	2012/06/06
		Fix bugs in deployVM window.
		Code refactoring.

1.3.0 	2012/05/28
		improved UI

1.2.2	2012/05/16
		Changed data structure.

1.2.1	2012/05/04
		Optimized error messages when Internet connection occured.
		Implemented folowing function,
		-edit original cloud information
		
0.040	2012/04/28
		Implemented following functions,
		-Deploy VM
		-Destroy VM

0.03	2012/04
		1st release.

Resources/myapp/ui/vm.js
(function() {
	myapp.cmd.vm = {};
	Titanium.include("../subroutine.js");
			
    myapp.cmd.vm.getData = function(json){
		var data =[];
    	for (var i=0; i<json.listvirtualmachinesresponse.virtualmachine.length;i++) {
     		var info = json.listvirtualmachinesresponse.virtualmachine[i];
 
 			if(info.state.match(/Running/)){
 				imgCircle = 'img/circle_blue.png';
 			}
 			else if(info.state.match(/Stopped/)){
 				imgCircle = 'img/circle_red.png';
 			}
 			else if(info.state.match(/Destroyed/)){
 				imgCircle = 'img/circle_gray.png';
 			}
 			else{
 				imgCircle = 'img/circle_yellow.png';
 			}

      		var circle = Ti.UI.createImageView({
      			image:imgCircle,
      			left:3,
     			height:30,
     			width:30
      		});      			
      		var labelId = Ti.UI.createLabel({
        		text:info.id,
        		font:{fontSize:10,fontWeight:'bold'}, textAlign:'left',
       			color:'#000',top:0, height:60, left:35, width:45
       		});
      		var labelDisplayName = Ti.UI.createLabel({
        		text:info.displayname,
        		font:{fontSize:10,fontWeight:'bold'}, textAlign:'left',
        		color:'#000',top:0, height:60, left:85, width:80
      		});
      		var labelTemplateName = Ti.UI.createLabel({
        		text:info.templatename,
        		font:{fontSize:10}, textAlign:'left',
        		color:'#000',top:0, height:60, left:170, width:'auto'
      		});
      		var row = Ti.UI.createTableViewRow({
        		className:"NomalCell",
        		height:60
      		});
 
 			row.add(circle);
      		row.add(labelId);
      		row.add(labelDisplayName);
      		row.add(labelTemplateName);
      		row.setHasChild(true);
      		data.push(row);
    	}
    	return data;
    };	
	myapp.cmd.vm.createWin = function(cloudName,apiUrl,apiKey,secretKey,opt){
		var cmd = "listVirtualMachines";
		var idZone;
		var nameZone;

		var addBtn = Titanium.UI.createButton({
			systemButton:Titanium.UI.iPhone.SystemButton.ADD
		});
		addBtn.addEventListener('click', function(){
			var list = [];
			var cmd = "listZones";
			var url = myapp.sub.getUrl(apiUrl,apiKey,secretKey,cmd,opt);
			var json;
			
	  		var xhr = Ti.Network.createHTTPClient();
  			xhr.open('GET', url, false);
  			xhr.onload = function() {
    			json = JSON.parse(this.responseText);
				
				var j=0;
    			for (var i=0; i<json.listzonesresponse.zone.length;i++) {
    				if((url.match(/datapipe\.com/) && json.listzonesresponse.zone[i].securitygroupsenabled) || !url.match(/datapipe\.com/)){
	      				list[j] = i;
	      				j++;
	      			}
				} 
				
				var w1 = Titanium.UI.createWindow({
					backgroundColor:'black',
					title:'Select Zone',
					barColor:'black',
				});
				var cancelBtn = Titanium.UI.createButton({
					systemButton:Titanium.UI.iPhone.SystemButton.CANCEL
				});

				w1.setLeftNavButton(cancelBtn);
				cancelBtn.addEventListener('click',function(){
					w1.close();
				});
			
				var nextBtn = Titanium.UI.createButton({
					title:'Next',
					width:100,
					height:40,
					top:300
				});
				nextBtn.addEventListener('click', function(){
					var list = [];
					var cmd = "listTemplates";
					var opt2 = opt + "&templatefilter=featured&zoneid=" + idZone;
					var url = myapp.sub.getUrl(apiUrl,apiKey,secretKey,cmd,opt2);
					var json;
					var idTemplate;
					var nameTemplate;
					
	  				var xhr = Ti.Network.createHTTPClient();
  					xhr.open('GET', url, false);
  					xhr.onload = function() {
    					json = JSON.parse(this.responseText);
						var w2 = Titanium.UI.createWindow({
							backgroundColor:'black',
							title:'Select Template',
							barColor:'black',
						});
			
						var nextBtn = Titanium.UI.createButton({
							title:'Next',
							width:100,
							height:40,
							top:300
						});
						nextBtn.addEventListener('click', function(){
							var list = [];
							var cmd = "listServiceOfferings";
							var url = myapp.sub.getUrl(apiUrl,apiKey,secretKey,cmd,opt);
							var json;
							var idServiceOffering;
							var nameServiceOffering;
			
	  						var xhr = Ti.Network.createHTTPClient();
  							xhr.open('GET', url, false);
  							xhr.onload = function() {
    							json = JSON.parse(this.responseText);
								var w3 = Titanium.UI.createWindow({
									backgroundColor:'black',
									title:'Select ServiceOffering',
									barColor:'black',
								});
			
								var nextBtn = Titanium.UI.createButton({
									title:'Next',
									width:100,
									height:40,
									top:300
								});
								nextBtn.addEventListener('click', function(){
									var related;
									var w4 = Ti.UI.createWindow({ 
										backgroundColor:'black',
										title:'Confirmation',
										barColor:'black', 
									});
		
									var label1 = Ti.UI.createLabel({
										text:nameZone,
										font:{fontSize:20, fontWeight:'bold'},
										textAlign:'center',
										color:'white',
										top:30,
										height:'auto',
										width:'auto'
									});
									w4.add(label1);
									
									var label2 = Ti.UI.createLabel({
										text:nameTemplate,
										font:{fontSize:20, fontWeight:'bold'},
										textAlign:'center',
										color:'white',
										top:90,
										height:'auto',
										width:'auto'
									});
									w4.add(label2);
									
									var label3 = Ti.UI.createLabel({
										text:nameServiceOffering,
										font:{fontSize:20, fontWeight:'bold'},
										textAlign:'center',
										color:'white',
										top:150,
										height:'auto',
										width:'auto'
									});
									w4.add(label3);

									var deployBtn = Titanium.UI.createButton({
										title:'Deploy',
										width:100,
										height:40,
										top:300
									});
									deployBtn.addEventListener('click', function(){
										var winArray = [w4,w3,w2,w1];
										if(apiUrl.match(/tatacommunications\.com/)){
											var cmd = "listNetworks";
											var url = myapp.sub.getUrl(apiUrl,apiKey,secretKey,cmd,opt);
											var json;
											
		 									var xhr = Ti.Network.createHTTPClient();
  											xhr.open('GET', url, false);
  											xhr.onload = function() {
    											json = JSON.parse(this.responseText);
    											related = String(json.listnetworksresponse.network[0].related);
  
  	  											var opt2 = opt + '&serviceofferingid=' + idServiceOffering + '&templateid=' + idTemplate + '&zoneid=' + idZone + '&networkIds=' + related;
	  											myapp.cmd.vm.deployWin(apiUrl,apiKey,secretKey,opt2,winArray);				
											}
		  									xhr.onerror = function(e){
    		    								alert(e);
       										};		
	  										xhr.send();	
										}
										else{
											var opt2 = opt + '&serviceofferingid=' + idServiceOffering + '&templateid=' + idTemplate + '&zoneid=' + idZone;
	  										myapp.cmd.vm.deployWin(apiUrl,apiKey,secretKey,opt2,winArray);			
										}								
									});
									w4.add(deployBtn);
									tabGroup.activeTab.open(w4,{animated:true});
								});
								w3.add(nextBtn);

								var picker = Ti.UI.createPicker();
								var data = [];
								for(var i=0;i<json.listserviceofferingsresponse.count;i++){
									data[i]=Ti.UI.createPickerRow({title:json.listserviceofferingsresponse.serviceoffering[i].name});
								}
			
								picker.selectionIndicator = true;
								picker.add(data);
								w3.add(picker);
								picker.setSelectedRow(0,1,true);
								idServiceOffering = json.listserviceofferingsresponse.serviceoffering[0].id;
								nameServiceOffering = json.listserviceofferingsresponse.serviceoffering[0].name;
				
								picker.addEventListener('change',function(e){
									idServiceOffering = json.listserviceofferingsresponse.serviceoffering[e.rowIndex].id;
									nameServiceOffering = json.listserviceofferingsresponse.serviceoffering[e.rowIndex].name;
								});
								tabGroup.activeTab.open(w3,{animated:true});
    						}
	  						xhr.onerror = function(e){
    	    					alert(e);
       						};		
	  						xhr.send();			
						});
						w2.add(nextBtn);

						var picker = Ti.UI.createPicker();
						var data = [];
						for(var i=0;i<json.listtemplatesresponse.count;i++){
							data[i]=Ti.UI.createPickerRow({title:json.listtemplatesresponse.template[i].name});
						}
			
						picker.selectionIndicator = true;
						picker.add(data);
						w2.add(picker);
						picker.setSelectedRow(0,1,true);
						idTemplate = json.listtemplatesresponse.template[0].id;
						nameTemplate = json.listtemplatesresponse.template[0].name;
				
						picker.addEventListener('change',function(e){
							idTemplate = json.listtemplatesresponse.template[e.rowIndex].id;
							nameTemplate = json.listtemplatesresponse.template[e.rowIndex].name;
						});
						tabGroup.activeTab.open(w2,{animated:true});
    				}
	  				xhr.onerror = function(e){
    	    			alert(e);
       				};		
	  				xhr.send();
				});
				w1.add(nextBtn);
			
				var picker = Ti.UI.createPicker();
				var data = [];
				for(var i=0;i<j;i++){
					data[i]=Ti.UI.createPickerRow({title:json.listzonesresponse.zone[list[i]].name});
				}
			
				picker.selectionIndicator = true;
				picker.add(data);
				w1.add(picker);
				picker.setSelectedRow(0,1,true);
				idZone = json.listzonesresponse.zone[list[0]].id;
				nameZone = json.listzonesresponse.zone[list[0]].name;
				
				picker.addEventListener('change',function(e){
					idZone = json.listzonesresponse.zone[list[e.rowIndex]].id;
					nameZone = json.listzonesresponse.zone[list[e.rowIndex]].name;
				});
				tabGroup.activeTab.open(w1,{animated:true});
    		}
	  		xhr.onerror = function(e){
    	    	alert(e);
       		};
       		
  			xhr.send();
		});
		
		var win = Ti.UI.createWindow({  
    		title:cloudName + ':VM',
    		backgroundColor:'white',
    		barColor:'black',
    		rightNavButton:addBtn
		});
		
		var url = myapp.sub.getUrl(apiUrl,apiKey,secretKey,cmd,opt);
		
  		var xhr = Ti.Network.createHTTPClient();
  		xhr.open('GET', url);
  		xhr.onload = function() {
    		var json = JSON.parse(this.responseText);
    		var data = myapp.cmd.vm.getData(json);
    		var tableView =Ti.UI.createTableView({
      			data:data,
      			editable:true
    		});   		    		
       		var r = myapp.sub.makeTableHeader();
       		var pulling = false;
			var reloading = false;
			tableView.headerPullView = r.tableHeader;	
			win.add(tableView);
    	   	tableView.addEventListener('scroll',function(e){
				var offset = e.contentOffset.y;
				if (offset <= -65.0 && !pulling){
					var t = Ti.UI.create2DMatrix();
					t = t.rotate(-180);
					pulling = true;
					r.arrow.animate({transform:t,duration:180});
					r.statusLabel.text = "Release to refresh...";
				}
				else if (pulling && offset > -65.0 && offset < 0){
					pulling = false;
					var t = Ti.UI.create2DMatrix();
					r.arrow.animate({transform:t,duration:180});
					r.statusLabel.text = "Pull down to refresh...";
				}
			});
			tableView.addEventListener('scrollEnd',function(e){
				if (pulling && !reloading && e.contentOffset.y <= -65.0){
					reloading = true;
					pulling = false;
					r.arrow.hide();
					r.actInd.show();
					r.statusLabel.text = "Reloading...";
					tableView.setContentInsets({top:60},{animated:true});
					r.arrow.transform=Ti.UI.create2DMatrix();
					var cmd = "listVirtualMachines";
					var url = myapp.sub.getUrl(apiUrl,apiKey,secretKey,cmd,opt);
					
			  		var xhr = Ti.Network.createHTTPClient();
  					xhr.open('GET', url);
  					xhr.onload = function() {
    					var json = JSON.parse(this.responseText);
    					var data = myapp.cmd.vm.getData(json);
						tableView.setContentInsets({top:0},{animated:true});
						reloading = false;
						r.statusLabel.text = "Pull down to refresh...";
						r.actInd.hide();
						r.arrow.show();
						tableView.data = data;
					}
					xhr.onerror = function(e){
    					alert(e);
       				};		
	  				xhr.send();	
				}
			});  	
			       		
			tableView.addEventListener('delete', function(e){
				var cmd = "destroyVirtualMachine";
				var opt2 = opt + '&id=' + Number(json.listvirtualmachinesresponse.virtualmachine[e.index].id);
				var url = myapp.sub.getUrl(apiUrl,apiKey,secretKey,cmd,opt2);
				var json2;
										
	  			var xhr = Ti.Network.createHTTPClient();
  				xhr.open('GET', url, false);
  				xhr.onload = function() {
    				json2 = JSON.parse(this.responseText);
					alert("VM ID:" + json.listvirtualmachinesresponse.virtualmachine[e.index].id + " has been destroyed.");
    			}
	  			xhr.onerror = function(e){
    	    		alert(e);
       			};		
	  			xhr.send();	
			});
				
		    tableView.addEventListener('click', function(e){
    			var detailLabel = Ti.UI.createLabel({
					color:'#999',
					text:"id:" + json.listvirtualmachinesresponse.virtualmachine[e.index].id + "\n" +
						"name:" + json.listvirtualmachinesresponse.virtualmachine[e.index].name + "\n" +
						"hostname:" + json.listvirtualmachinesresponse.virtualmachine[e.index].hostname + "\n" +
						"displayname:" + json.listvirtualmachinesresponse.virtualmachine[e.index].displayname + "\n" +
						"group:" + json.listvirtualmachinesresponse.virtualmachine[e.index].group + "\n" +
						"state:" + json.listvirtualmachinesresponse.virtualmachine[e.index].state + "\n" +
						"templatename:" + json.listvirtualmachinesresponse.virtualmachine[e.index].templatename + "\n" +
						"cpunumber:" + json.listvirtualmachinesresponse.virtualmachine[e.index].cpunumber + "\n" +
						"cpuspeed:" + json.listvirtualmachinesresponse.virtualmachine[e.index].cpuspeed + "\n" +
						"memory:" + json.listvirtualmachinesresponse.virtualmachine[e.index].memory + "\n" +
						"zonename:" + json.listvirtualmachinesresponse.virtualmachine[e.index].zonename,
					font:{fontSize:14,fontFamily:'Helvetica Neue'},
					textAlign:'left',
					top:10,
					left:5,
					height:'auto',
					width:'auto'
				});
				
				var detailWin = Ti.UI.createWindow({
					title: cloudName + ':VM:' + json.listvirtualmachinesresponse.virtualmachine[e.index].id,
					backgroundColor: 'white',
					barColor:'black'
				});
		
				detailWin.add(detailLabel);
				tabGroup.activeTab.open(detailWin,{animated:true});
			});
  		};
  		xhr.send();
  		xhr.onerror = function(){
        	alert("No Internet connection.Please make sure that you have Internet connectivity and try again later.");
       	};
  		return win;
 	};
 	
 	myapp.cmd.vm.deployWin = function(apiUrl,apiKey,secretKey,opt,winArray){
 		var cmd = "deployVirtualMachine";
		var url = myapp.sub.getUrl(apiUrl,apiKey,secretKey,cmd,opt);
		var json;
										
	  	var xhr = Ti.Network.createHTTPClient();
  		xhr.open('GET', url, false);
  		xhr.onload = function() {
    		json = JSON.parse(this.responseText);
			var w5 = Ti.UI.createWindow({ 
				backgroundColor:'black',
				title:'Started to deploy',
				barColor:'black', 
			});
			var label1 = Ti.UI.createLabel({
				text:'Please wait a few minutes.',
				font:{fontSize:20, fontWeight:'bold'},
				textAlign:'center',
				color:'white',
				top:30,
				height:'auto',
				width:'auto'
			});
			w5.add(label1);
											
			var okBtn = Titanium.UI.createButton({
				title:'OK',
				width:100,
				height:40,
				top:300
			});
			w5.add(okBtn);
			tabGroup.activeTab.open(w5,{animated:true});

			okBtn.addEventListener('click', function(){
				tabGroup.activeTab.close(w5);
				for(var i=0;i<winArray.length;i++){
					tabGroup.activeTab.close(winArray[i]);
				}
			});
    	}
    									
	  	xhr.onerror = function(e){
    		alert(e);
       	};		
	  	xhr.send();	
 	};
})();

Resources/myapp/ui/map.js
(function() {
	myapp.map = {};
	myapp.map.createTab = function(){
		var win = Ti.UI.createWindow({
			barColor:'black',
			title: 'Map',
			});
		var tab = Ti.UI.createTab({  
			title:'Map',
    		icon:'img/light_globe.png',
    		window:win
		});
		
		var mapview = Titanium.Map.createView({
    		mapType: Titanium.Map.STANDARD_TYPE,
    		region:{latitude:40, longitude:130, latitudeDelta:90, longitudeDelta:360},
    		animate:true,
    		regionFit:false,
		});

		var Contegix = Titanium.Map.createAnnotation({
        	latitude:38.625542,
        	longitude:-90.194946,
        	title:"Contegix",
        	subtitle:'Saint Louis',
        	pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var Datapipe1 = Titanium.Map.createAnnotation({
        	latitude:40.745491,
        	longitude:-73.982758,
        	title:"Datapipe",
        	subtitle:'New York',
        	pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var Datapipe2 = Titanium.Map.createAnnotation({
	        latitude:22.284881,
    	    longitude:114.154996,
        	title:"Datapipe",
	        subtitle:'Hong Kong',
    	    pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var Datapipe3 = Titanium.Map.createAnnotation({
	        latitude:37.378615,
    	    longitude:-121.916656,
        	title:"Datapipe",
	        subtitle:'Silicon Valley',
    	    pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var Datapipe4 = Titanium.Map.createAnnotation({
        	latitude:51.518593,
	        longitude:-0.098476,
    	    title:"Datapipe",
        	subtitle:'London',
	        pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var Datapipe5 = Titanium.Map.createAnnotation({
    		    latitude:31.165572,
        		longitude:121.49398,
		        title:"Datapipe",
    		    subtitle:'Shanghai',
	    	    pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var IDCF = Titanium.Map.createAnnotation({
    	    latitude:35.689488,
        	longitude:139.691706,
	        title:"IDCF",
    	    subtitle:'Tokyo',
        	pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var NineFold = Titanium.Map.createAnnotation({
        	latitude:-33.871584,
	        longitude:151.203054,
    	    title:"NineFold",
        	subtitle:'Sydney',
	        pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var NTTCom2 = Titanium.Map.createAnnotation({
    	    latitude:37.354108,
        	longitude:-121.955236,
        	title:"NTTCom",
        	subtitle:'USA',
        	pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var Tata1 = Titanium.Map.createAnnotation({
    	    latitude:19.075984,
        	longitude:72.877656,
	        title:"Tata",
    	    subtitle:'South India',
        	pincolor:Titanium.Map.ANNOTATION_GREEN
		});
		var Tata2 = Titanium.Map.createAnnotation({
        	latitude:1.335503,
        	longitude:103.890461,
	        title:"Tata",
    	    subtitle:'Singapore',
        	pincolor:Titanium.Map.ANNOTATION_GREEN
		});

		// ピンの追加
		mapview.addAnnotations([Contegix, Datapipe1, Datapipe2, Datapipe3, Datapipe4, Datapipe5, IDCF, NineFold, NTTCom2, Tata1, Tata2]);
		win.add(mapview);

/*
// GPS の利用目的を表示。iOS 4 以降では必須。
Titanium.Geolocation.purpose = 'サンプル';

Titanium.Geolocation.getCurrentPosition(
    function(e) {
        if (!e.success || e.error){
            alert('位置情報が取得できませんでした');
            return;
        }

        latitude = e.coords.latitude;
        longitude = e.coords.longitude;
        Ti.API.info('LAT:' + latitude);
        Ti.API.info('LONG:' + longitude);
        
        // ピンを立てる
        var currentPos = Titanium.Map.createAnnotation({
            latitude:latitude,
            longitude:longitude,
            title:"現在地",
            pincolor:Titanium.Map.ANNOTATION_GREEN,
            animate:true
        });
        mapview.addAnnotation(currentPos);
        mapview.show(); // 隠していた地図を表示する
        mapview.setLocation({   // 現在地まで地図をスクロールする
            latitude:latitude,
            longitude:longitude,
            latitudeDelta:0.01,
            longitudeDelta:0.01
        });
    }
);
*/

		return tab;	
	}
})();

Resources/myapp/ui/tab.js
(function() {
	myapp.tab = {};
	Titanium.include("../subroutine.js");
	Titanium.include("vm.js");
	Titanium.include("volume.js");
		
	myapp.tab.beginReloading = function(data, count, apiUrl, apiKey, secretKey, cmdLists,opt, cloudName){
		if(!apiKey || !secretKey || !apiKey[cloudName] || !secretKey[cloudName]){
			for(var cmd in cmdLists){
				count[cmd] = 'NA';
			}
		}
		else{	
			for(var cmd in cmdLists){			
				var url = myapp.sub.getUrl(apiUrl,apiKey[cloudName],secretKey[cloudName],cmdLists[cmd],opt);
  				// オブジェクトを生成します。
  				var xhr = Ti.Network.createHTTPClient();
  				xhr.open('GET', url, false);

  				// データダウンロード時のイベント処理
  				xhr.onload = function() {
    				var json = JSON.parse(this.responseText);
    				//i.API.info(json);
    				if(this.responseText.match(/listvirtualmachinesresponse/)){
	    				count['VM'] = json.listvirtualmachinesresponse.count;
	    				if(!count['VM']){
	    					count['VM'] = 0;
	    				}
	    			}
	    			else if(this.responseText.match(/listvolumesresponse/)){
	    				count['VOLUME'] = json.listvolumesresponse.count;
	    				if(!count['VOLUME']){
	    					count['VOLUME'] = 0;
	    				}
	    			}
				};				
  				xhr.onerror = function(event){
       				//alert(event);
       			};
       			
        		// HTTPリクエストの送信
  	 			xhr.send();
			}
			
		}
		data[0].title = 'VM(' + count['VM'] +')';
		data[1].title = 'VOLUME(' + count['VOLUME'] +')';
		return data;
	};

	
	myapp.tab.createTab = function(cloudName,flag,apiUrl,opt){
		var apiKey = new Array(); 
		var secretKey = new Array();
		var cmdLists = {
			'VM':'listVirtualMachines', 
			'VOLUME':'listVolumes'
			};
		var count = new Object();
	
		var btn = Ti.UI.createButton({
			backgroundImage:flag,
			height:22,
			width:32
		});
		
		var win = Ti.UI.createWindow({  
    		title:cloudName,
    		backgroundColor:'white',
    		barColor:'black',
       		rightNavButton:btn
		});

		win.addEventListener('open', function(){
			if(Ti.App.Properties.getString('apiKey')){
				apiKey = JSON.parse(Ti.App.Properties.getString('apiKey'));
			}
			if(Ti.App.Properties.getString('secretKey')){
				secretKey = JSON.parse(Ti.App.Properties.getString('secretKey'));
			}
			
			data = myapp.tab.beginReloading(data, count, apiUrl, apiKey, secretKey, cmdLists,opt, cloudName);
			tableView.data = data;
		});
		
		var tab = Ti.UI.createTab({  
    		icon:'img/light_globe.png',
    		title:cloudName,
    		window:win
		});
		
		var border = Ti.UI.createView({
			backgroundColor:"#576c89",
			height:2,
			bottom:0
		});

		var tableHeader = Ti.UI.createView({
			backgroundColor:"#e2e7ed",
			width:320,
			height:60
		});
		tableHeader.add(border);

		var arrow = Ti.UI.createView({
			backgroundImage:"img/whiteArrow.png",
			width:23,
			height:60,
			bottom:10,
			left:20
		});

		var statusLabel = Ti.UI.createLabel({
			text:"Pull to reload",
			left:55,
			width:200,
			bottom:30,
			height:"auto",
			color:"#576c89",
			textAlign:"center",
			font:{fontSize:13,fontWeight:"bold"},
			shadowColor:"#999",
			shadowOffset:{x:0,y:1}
		});
		
		var actInd = Titanium.UI.createActivityIndicator({
			left:20,
			bottom:13,
			width:30,
			height:30
		});

		tableHeader.add(arrow);
		tableHeader.add(statusLabel);
		tableHeader.add(actInd);

		var pulling = false;
		var reloading = false;

		var data = [
			{title:'VM(' + count['VM'] +')', hasChild:true, leftImage:'img/Crystal_Clear_app_network_local.png', height:Ti.UI.FILL},
			{title:'VOLUME(' + count['VOLUME'] +')', hasChild:true, leftImage:'img/Crystal_Clear_app_database.png', height:Ti.UI.FILL}
	
		];
		
		// TableViewの作成
    	var tableView =Ti.UI.createTableView({
      		data:data
    	});

		tableView.headerPullView = tableHeader;			
    	win.add(tableView);

		apiKey = JSON.parse(Ti.App.Properties.getString('apiKey'));
		secretKey = JSON.parse(Ti.App.Properties.getString('secretKey'));
		
		data = myapp.tab.beginReloading(data, count, apiUrl, apiKey, secretKey, cmdLists,opt, cloudName);
		tableView.data = data;
			
		tableView.addEventListener('click', function(event){
			if(Ti.App.Properties.getString('apiKey')){
				apiKey = JSON.parse(Ti.App.Properties.getString('apiKey'));
			}
			if(Ti.App.Properties.getString('secretKey')){
				secretKey = JSON.parse(Ti.App.Properties.getString('secretKey'));
			}
			
    		if(event.rowData.title.match(/VM/)){
  				if(apiKey[cloudName] && secretKey[cloudName]){
					var detailWin = myapp.tab.vm.createWin(cloudName,apiUrl,apiKey[cloudName],secretKey[cloudName],opt,tab);
					tab.open(detailWin);
				}
				else{
					alert("Please input API KEY & SECRET KEY.");
				}
			}
			else if(event.rowData.title.match(/VOLUME/)){
				if(apiKey[cloudName] && secretKey[cloudName]){
					var detailWin = myapp.tab.volume.createWin(cloudName,apiUrl,apiKey[cloudName],secretKey[cloudName],opt,tab);
					tab.open(detailWin);
				}
				else{
					alert("Please input API KEY & SECRET KEY.");
				}
			}
		});
       	tableView.addEventListener('scroll',function(event){
			var offset = event.contentOffset.y;
			if (offset <= -65.0 && !pulling){
				var t = Ti.UI.create2DMatrix();
				t = t.rotate(-180);
				pulling = true;
				arrow.animate({transform:t,duration:180});
				statusLabel.text = "Release to refresh...";
			}
			else if (pulling && offset > -65.0 && offset < 0){
				pulling = false;
				var t = Ti.UI.create2DMatrix();
				arrow.animate({transform:t,duration:180});
				statusLabel.text = "Pull down to refresh...";
			}
		});

		tableView.addEventListener('scrollEnd',function(event){
			if (pulling && !reloading && event.contentOffset.y <= -65.0){
				reloading = true;
				pulling = false;
				arrow.hide();
				actInd.show();
				statusLabel.text = "Reloading...";
				tableView.setContentInsets({top:60},{animated:true});
				arrow.transform=Ti.UI.create2DMatrix();
				data = myapp.tab.beginReloading(data, count, apiUrl, apiKey, secretKey, cmdLists,opt, cloudName);
				//setTimeout(myapp.tab.endReloading(tableView, statusLabel, actInd, arrow),2000);
			
				tableView.setContentInsets({top:0},{animated:true});
				reloading = false;
				statusLabel.text = "Pull down to refresh...";
				actInd.hide();
				arrow.show();
		
				tableView.data = data;
			}
		});  	
  		return tab;
 	};
})();



Resources/app.js
var myapp = {};

Titanium.include("lib/sha1.js");
Titanium.include("myapp/ui/view.js");
Titanium.include("myapp/ui/setting.js");
Titanium.include("myapp/ui/map.js");

var cmdLists = {
	'VM':{
		'list':'listVirtualMachines',
		'res':'listvirtualmachinesresponse',
		'key':'virtualmachine'
	}, 
	'VOLUME':{
		'list':'listVolumes',
		'res':'listvolumesresponse',
		'key':'volume'
	}
};	
		
var cloudOrig = [
	{title:"Contegix", img:"img/us.png", url:"https://cloud.contegix.com/portal/client/api", hasChild:true, height:Ti.UI.FILL},
	{title:"Datapipe", img:"img/us.png", url:"https://cloud.datapipe.com/api/compute/v1", hasChild:true, height:Ti.UI.FILL},
	{title:"IDCF", img:"img/jp.png", url:"https://api.noahcloud.jp/portal/client/api", hasChild:true, height:Ti.UI.FILL},
	{title:"NineFold", img:"img/au.png", url:"https://api.ninefold.com/compute/v1.0/", hasChild:true, height:Ti.UI.FILL},
	{title:"NTTCom", img:"img/jp.png", url:"https://mycloud2.securesites.com/client/api", hasChild:true, height:Ti.UI.FILL},
	{title:"Tata", img:"img/in.png", url:"http://manage.iaas.tatacommunications.com:8080/client/api", hasChild:true, height:Ti.UI.FILL}
];
	
var cloud = new Array();
if(Ti.App.Properties.getString('cloud')){
	cloud = JSON.parse(Ti.App.Properties.getString('cloud'));
}
else{
	cloud = cloudOrig.slice(0);
}

var cloudLen = cloud.length;
var opt = "response=json";


Ti.UI.setBackgroundColor('#000');

var tabGroup = Ti.UI.createTabGroup();
var tabSetting = myapp.setting.createTab();
var tabMap = myapp.map.createTab();

var viewArray = new Array();
var btn = new Array();
for(var i=0;i<cloud.length;i++){
	viewArray[i] = myapp.view.createView(cloud[i]['title'],cloud[i]['img'],cloud[i]['url'],opt);
}

var i=0;		
var cloudWin = Ti.UI.createWindow({
	backgroundColor:'black',
	barColor:'black',
	title:cloud[i]['title'],
	rightNavButton:Ti.UI.createButton({backgroundImage:cloud[i]['img'], height:22, width:32})
});

var scrollView = Titanium.UI.createScrollableView({
    views:viewArray,
    showPagingControl:true,
    pagingControlHeight:30,
    pagingControlColor: 'black',
    maxZoomScale:0,
    currentPage:0
});

var activeView = viewArray[0];
scrollView.addEventListener('scroll', function(e){
	activeView = e.view;  
	i = e.currentPage;
	if(i >= 0){
		cloudWin.setTitle(cloud[i]['title']);
		cloudWin.setRightNavButton(Ti.UI.createButton({backgroundImage:cloud[i]['img'], height:22, width:32}));
	}
});
cloudWin.add(scrollView);

var tabCloud = Ti.UI.createTab({
	title:'Cloud',
	icon:'img/Cloud.png',
	window:cloudWin
});

tabGroup.addTab(tabCloud);
tabGroup.addTab(tabMap); 
tabGroup.addTab(tabSetting); 

if(Ti.App.Properties.getString('addCloud')){
	addCloud = JSON.parse(Ti.App.Properties.getString('addCloud'));
	cloud[cloudLen] = addCloud;
	viewadd = myapp.view.createView(cloud[cloudLen]['title'],cloud[cloudLen]['img'],cloud[cloudLen]['url'],opt);
	scrollView.addView(viewadd);
}

tabGroup.open();